hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/master/BackupLogCleaner.java

```diff
@@ -81,39 +81,55 @@ public void init(Map<String, Object> params) {
     }
   }
 
-  private Map<Address, Long> getServerToNewestBackupTs(List<BackupInfo> backups)
+  /**
+   * Calculates the timestamp boundary up to which all backup roots have already included the WAL.
+   * I.e. WALs with a lower (= older) or equal timestamp are no longer needed for future incremental
+   * backups.
+   */
+  private Map<Address, Long> serverToPreservationBoundaryTs(List<BackupInfo> backups)
```

Comment:
should this method have a precondition asserting that all BackupInfos in the list have an identical table set?

Otherwise I think this could cause data loss in an incremental backup. Let's say we have one-table and another-table on our cluster, and we back up each one independently. Let's say we first took a backup for one-table, then a backup for another-table. Then we passed the corresponding BackupInfos into this method, this would yield preservation boundaries which would remove WALs necessary for the next incremental backup of one-table

Reply:
I don't see an issue. Trying to follow your example:

- Imagine 3 servers S1, S2, S3
- table T1 has regions on S1 and S2, table T2 has regions on S2 and S3
- we have 2 backup roots R1 that backs up T1, and R2 that backs up T2

At time:
- t=0, we backup T1 in R1 => backup B1
- t=0, we backup T2 in R2 => backup B2
- t=10, we backup T1 in R1 => backup B3
- t=20, we backup T2 in R2. => backup B4

Following the logic in this method:
- newestBackupPerRootDir will contain: (R1: B3, R2: B4)
- boundaries will contain: (S1: 10, S2: 10, S3: 20)

So for T1, all WALs up to t=10 can be deleted, for T2, WALs will be preserved from t=10 or t=20, depending whether other tables are present.

Reply:
Thank you for the explanation, both this and below. I think I follow you. Let me rephrase.

`newestBackupPerRootDir` is a "race to the top" in terms of identifying the most recent timestamp across backups, while `boundaries` is a "race to the bottom" in terms of identifying the least recent backup present on each server. By selecting the "oldest of the newest" you determine the minimum timestamp that must be preserved for each server. Any WALs older than this minimum timestamp can be safely discarded.

Reply:
Cool, I think the missing piece for me was that a backup root was specific to a given table
> we have 2 backup roots R1 that backs up T1, and R2 that backs up T2

Reply:
> Cool, I think the missing piece for me was that a backup root was specific to a given table

It isn't. Each backup root contains one or more tables, i.e. a set of tables, these table sets may or may not have tables in common.

Reply:
Hmm, okay so reframing your example in a single root:

Imagine 
3 servers S1, S2, S3
table T1 has regions on S1 and S2, table T2 has regions on S2 and S3 
we have 1 backup root R1 that backs up T1 and T2

At time:
t=0, we backup T1 in R1 => backup B1
t=0, we backup T2 in R1 => backup B2
t=10, we backup T1 in R1 => backup B3
t=20, we backup T2 in R1 => backup B4

Following the logic in this method:

`newestBackupPerRootDir` will contain: (R1: B4)

We would calculate the preservation boundaries here:
```java
    // This map tracks, for every address, the least recent (= oldest / lowest timestamp) inclusion
    // in any backup. In other words, it is the timestamp boundary up to which all backups roots
    // have included the WAL in their backup.
    Map<Address, Long> boundaries = new HashMap<>();
    for (BackupInfo backupInfo : newestBackupPerRootDir.values()) {
      for (TableName table : backupInfo.getTables()) {
        for (Map.Entry<String, Long> entry : backupInfo.getTableSetTimestampMap().get(table)
          .entrySet()) {
          Address address = Address.fromString(entry.getKey());
          Long storedTs = boundaries.get(address);
          if (storedTs == null || entry.getValue() < storedTs) {
            boundaries.put(address, entry.getValue());
          }
        }
      }
    }
```

Since the newest backup in R1 is B4, `serverToPreservationBoundaryTs` will contain (S1: 20, S2: 20, S3: 20)

In this situation, we must not delete WALs from S1 or S2 between times 10 and 20 because a subsequent incremental backup of T1 will require those WALs.

In the BackupLogCleaner, though, we will end up calling `canDeleteFile` and hitting this code with that `serverToPreservationBoundaryTs` (S1: 20, S2: 20, S3: 20) renamed as `addressToBoundaryTs`:

```java
      if (!addressToBoundaryTs.containsKey(walServerAddress)) { // for S1 & S2, this will be false
         // irrelevant ...
      }

      Long backupBoundary = addressToBoundaryTs.get(walServerAddress); // for S1 & S2, this will be 20
      if (backupBoundary >= walTimestamp) { // For WALs between times 10 and 20, this will be true
        if (LOG.isDebugEnabled()) {
          LOG.debug(
            "Backup found for server: {}. All backups from {} are newer than file, so deleting: {}",
            walServerAddress.getHostName(), backupBoundary, path);
        }
        // so we will erroneously delete WALs <20, despite the next inc backup of T1 requiring them
        return true;
      }
```

I guess this depends on how ancestry is defined? If we consider all backups in a root to be ancestors regardless of their table set, then maybe it is okay to delete these WALs. But, if not, then I don't see how it is okay?

Reply:
> newestBackupPerRootDir will contain: (R1: B4)

Agreed

> Since the newest backup in R1 is B4, serverToPreservationBoundaryTs will contain (S1: 20, S2: 20, S3: 20)

I think this is wrong.

To make things more concrete, I'm going to assume all backups were full backups, and B1 == B2 (since it has the same timestamp). I.e. the backups were:
- B1: at timestamp 0, containing tables T1 & T2
- B3: at timestamp 10, containing T1
- B4: at timestamp 20, containing T2

At this point, the data in `backupInfo.getTableSetTimestampMap()` will be:
```
T1:
  S1: 10
  S2: 10
T2:
  S2: 20
  S3: 20 
```

and `serverToPreservationBoundaryTs` will be (S1: 10, S2: 10, S3: 20).

The reason that the BackupInfo of B4 contains the log timestamps of tables not included in B4 is due to how the backup client updates these:
```
// From: FullTableBackupClient

// Updates only the rows for the tables included in B4
backupManager.writeRegionServerLogTimestamp(backupInfo.getTables(), newTimestamps);

// Reads the rows for all tables once included in a backup in this backup root
Map<TableName, Map<String, Long>> newTableSetTimestampMap = backupManager.readLogTimestampMap();
```

Reply:
Sorry for being a pain here, but I'm not sure I agree. When building `serverToPreservationBoundaryTs` we loop through:
```
for (BackupInfo backupInfo : newestBackupPerRootDir.values()) {
  // build boundaries via tableSetTimestampMap, like you said
}
```
and we agree that `newestBackupPerRootDir.values()` will only contain B4. So our boundaries would end up only being based on the timestamps from B4? How does the second newest backup, and beyond, in the root come into play?

Reply:
Answering my own questions a bit here, I think I needed to look deeper into how the BackupInfo gets constructed. When the tableSetTimestampMap is [populated](https://github.com/apache/hbase/blob/78da9e962b56846cea1db6dd43f9cd02c632ecf0/hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupSystemTable.java#L926-L966), **it will contain entries for every table in the root**, not just every table in the given backup. So while B4 might only pertain to T2, its corresponding BackupInfo will be aware of the timestamps for T1. That's a confusing design imo, but if  true then I think this logic is sound

Reply:
It's indeed like that. Doing it that way is a good way to avoid some special cases where backups in the same root backup do not contain the same tables, but the docs surrounding those BackupInfo concepts is indeed lacking.